File: frontend/src/lib/unitConversion.ts
// Mirror of backend/app/services/unit_conversion.py — keep the two in sync.

export const MEAL_UNITS = ['piece', 'cup', 'tbsp', 'tsp', 'gram', 'ml', 'glass'] as const
export type MealUnit = (typeof MEAL_UNITS)[number]

// Approximate grams per 1 unit. "piece" is derived per-item (estimated_grams / quantity).
export const UNIT_GRAMS: Partial<Record<string, number>> = {
  cup: 240,
  glass: 250,
  tbsp: 15,
  tsp: 5,
  gram: 1,
  ml: 1,
}

// Units measured in fluid volume — show their size in ml rather than g.
const LIQUID_UNITS = new Set(['glass', 'ml'])

/** Human label for a unit dropdown option, e.g. "cup (240g)", "glass (250ml)", "piece". */
export function unitLabel(unit: string): string {
  const grams = UNIT_GRAMS[unit]
  if (!grams) return unit
  if (unit === 'ml') return unit
  return `${unit} (${grams}${LIQUID_UNITS.has(unit) ? 'ml' : 'g'})`
}

export interface ScalableItem {
  quantity: number
  unit: string
  estimated_grams: number
  calories: number
  calorie_low: number
  calorie_high: number
  carbs_g: number
  protein_g: number
  fat_g: number
  fibre_g: number
  user_edited_fields: string[]
}

const SCALABLE_FIELDS = [
  'calories', 'calorie_low', 'calorie_high',
  'carbs_g', 'protein_g', 'fat_g', 'fibre_g', 'estimated_grams',
] as const

function gramsPerUnit(item: ScalableItem, unit: string): number | null {
  const table = UNIT_GRAMS[unit]
  if (table) return table
  if (item.quantity && item.estimated_grams) return item.estimated_grams / item.quantity
  return null
}

/** Rescale calories/macros proportionally on quantity/unit change,
 *  never touching fields the user has explicitly corrected. */
export function scaleItem<T extends ScalableItem>(item: T, newQuantity: number, newUnit?: string): T {
  const unit = newUnit || item.unit
  const edited = new Set(item.user_edited_fields)
  let factor: number
  if (unit === item.unit) {
    factor = item.quantity ? newQuantity / item.quantity : 1
  } else {
    const oldGpu = gramsPerUnit(item, item.unit)
    const newGpu = gramsPerUnit(item, unit)
    factor = oldGpu && newGpu && item.quantity
      ? (newQuantity * newGpu) / (item.quantity * oldGpu)
      : (item.quantity ? newQuantity / item.quantity : 1)
  }
  const scaled = { ...item, quantity: newQuantity, unit }
  for (const field of SCALABLE_FIELDS) {
    if (edited.has(field) || item[field] == null) continue
    scaled[field] = Math.round(item[field] * factor * 10) / 10
  }
  return scaled
}

File: backend/app/services/unit_conversion.py
"""Canonical unit -> grams table and proportional item scaling.

Single source of truth for household-unit conversions; mirrored client-side
in frontend/src/lib/unitConversion.ts for instant UI feedback.
"""
from typing import List, Optional

# Approximate grams per 1 unit for common Indian household measures.
# "piece" is intentionally absent: piece weight varies per food (idli vs dosa),
# so it is derived from the item's own estimated_grams / quantity.
UNIT_GRAMS = {
    "cup": 240,
    "glass": 250,
    "tbsp": 15,
    "tsp": 5,
    "gram": 1,
    "ml": 1,
}

SCALABLE_FIELDS = (
    "calories",
    "calorie_low",
    "calorie_high",
    "carbs_g",
    "protein_g",
    "fat_g",
    "fibre_g",
    "estimated_grams",
)


def _grams_per_unit(item: dict, unit: str) -> Optional[float]:
    if unit in UNIT_GRAMS:
        return float(UNIT_GRAMS[unit])
    # piece (or unknown unit): derive from the item's own estimate
    quantity = item.get("quantity") or 0
    grams = item.get("estimated_grams") or 0
    if quantity and grams:
        return grams / quantity
    return None


def scale_item(
    item: dict,
    new_quantity: float,
    new_unit: Optional[str] = None,
    user_edited_fields: Optional[List[str]] = None,
) -> dict:
    """Return a copy of item with macros/calories rescaled proportionally.

    Fields listed in user_edited_fields are left untouched (never silently
    overwrite a user's correction).
    """
    edited = set(user_edited_fields or [])
    old_quantity = item.get("quantity") or 1
    old_unit = item.get("unit") or "piece"
    new_unit = new_unit or old_unit

    if new_unit == old_unit:
        factor = new_quantity / old_quantity if old_quantity else 1.0
    else:
        old_gpu = _grams_per_unit(item, old_unit)
        new_gpu = _grams_per_unit(item, new_unit)
        if old_gpu and new_gpu and old_quantity:
            factor = (new_quantity * new_gpu) / (old_quantity * old_gpu)
        else:
            factor = new_quantity / old_quantity if old_quantity else 1.0

    scaled = dict(item)
    scaled["quantity"] = new_quantity
    scaled["unit"] = new_unit
    for field in SCALABLE_FIELDS:
        if field in edited or item.get(field) is None:
            continue
        scaled[field] = round(item[field] * factor, 1)
    return scaled
